cpp/src/arrow/util/logging.cc

```diff
@@ -113,10 +115,14 @@ void ArrowLog::StartArrowLog(const std::string& app_name,
                              ArrowLogLevel severity_threshold,
                              const std::string& log_dir) {
   severity_threshold_ = severity_threshold;
-  app_name_.reset(new std::string(app_name.c_str()));
+  // In InitGoogleLogging, it simply keeps the pointer.
+  // We need to make sure the app name passed to InitGoogleLogging exist.
+  // We should avoid using static string is a dynamic lib.
+  static std::unique_ptr<std::string> app_name_;
+  app_name_.reset(new std::string(app_name));
+  log_dir_.reset(new std::string(log_dir));
 #ifdef ARROW_USE_GLOG
   int mapped_severity_threshold = GetMappedSeverity(severity_threshold_);
-  google::InitGoogleLogging(app_name_->c_str());
```

Comment:
Is it required to move this below? Intuitively, I would expect the init call to have to occur first.

Side note: I haven't been able to find a documentation for the glog API. Is there any?

Reply:
This is not the document, but explains why we move this below: https://github.com/google/glog/blob/3267f3e1a8dd629922dbe28c09a5fa5e60de5a84/src/logging_unittest.err#L7 . In this function, we want to provide the ability that if the users don't specify the log dir, the only should only output stderr. Therefore, `google::InitGoogleLogging`  is called only when `log_dir_` is not empty.

Reply:
Ok, can you add comments explaining this?

Reply:
Added.